staggered contraction code

alexstrel opened the pull request:
This PR primarily adds an extension to the QUDA contraction code for staggered fermions. To utilize this, a new parameter, `QUDA_CONTRACT_TYPE_STAGGERED_FT_T`, has been introduced. 
The main application for testing this is contractFT_test.cpp.

cpviolator committed:
GK merge

AlexVaq committed:
First implementation of the shifts required for the staggered spin-taste operator.
Includes an interface for the covariant derivative (valid for any regularization).

AlexVaq committed:
First implementation of the shifts required for the staggered spin-taste operator.
Includes an interface for the covariant derivative (valid for any regularization).

AlexVaq committed:
First implementation of the shifts required for the staggered spin-taste operator.
Includes an interface for the covariant derivative (valid for any regularization).

AlexVaq committed:
First implementation of the shifts required for the staggered spin-taste operator.
Includes an interface for the covariant derivative (valid for any regularization).

AlexVaq committed:
First implementation of the shifts required for the staggered spin-taste operator.
Includes an interface for the covariant derivative (valid for any regularization).

AlexVaq committed:
First implementation of the shifts required for the staggered spin-taste operator.
Includes an interface for the covariant derivative (valid for any regularization).

AlexVaq committed:
Fixed typo in prototype input for qudaShift

AlexVaq committed:
Fixed typo in prototype input for qudaShift

AlexVaq committed:
Fixed typo in prototype input for qudaShift

AlexVaq committed:
Fixed typo in prototype input for qudaShift

AlexVaq committed:
Fixed typo in prototype input for qudaShift

AlexVaq committed:
Fixed typo in prototype input for qudaShift

AlexVaq committed:
Bug fixes

AlexVaq committed:
Bug fixes

AlexVaq committed:
Bug fixes

AlexVaq committed:
Bug fixes

AlexVaq committed:
Bug fixes

AlexVaq committed:
Bug fixes

AlexVaq committed:
Added covDev for staggered

AlexVaq committed:
Added covDev for staggered

AlexVaq committed:
Added covDev for staggered

AlexVaq committed:
Added covDev for staggered

AlexVaq committed:
Added covDev for staggered

AlexVaq committed:
Added covDev for staggered

AlexVaq committed:
Made the gauge field persistent

AlexVaq committed:
Made the gauge field persistent

AlexVaq committed:
Made the gauge field persistent

AlexVaq committed:
Made the gauge field persistent

AlexVaq committed:
Made the gauge field persistent

AlexVaq committed:
Made the gauge field persistent

AlexVaq committed:
Made the gauge field persistent

AlexVaq committed:
Made the gauge field persistent

AlexVaq committed:
Cleaning stuff up

AlexVaq committed:
Cleaning stuff up

AlexVaq committed:
Cleaning stuff up

AlexVaq committed:
Cleaning stuff up

AlexVaq committed:
Cleaning stuff up

AlexVaq committed:
Cleaning stuff up

AlexVaq committed:
Cleaning stuff up

AlexVaq committed:
Cleaning stuff up

cpviolator committed:
GK merge

cpviolator committed:
Merge remote-tracking branch 'origin/feature/generic_kernel' into feature/hotqcd_correlators_gk_merged

cpviolator committed:
Merge remote-tracking branch 'origin/feature/generic_kernel' into feature/hotqcd_correlators_gk_merged

AlexVaq committed:
Merge branch 'feature/spin-taste_operator' of github.com:lattice/quda into feature/spin-taste_operator

cpviolator committed:
Tweak parameter setting calls

cpviolator committed:
Tweak parameter setting calls

cpviolator committed:
Tweak parameter setting calls

Commit:
Merge branch 'feature/generic_kernel' into feature/spin-taste_operator

cpviolator committed:
Bring even with develop

cpviolator committed:
Apply new method of ReduceArg construction

Commit:
Merge branch 'feature/hotqcd_correlators_gk_merged' into try-merge
Branch try-merge is a trial merge of feature/hotqcd_correlators_gk_merged into branch feature/staggered_correlators_gk

Commit:
Resolve multiple definitions of blas test configuration parameters.

AlexVaq committed:
Spin-taste operator for MILC, v1

AlexVaq committed:
Spin-taste operator for MILC, v1

AlexVaq committed:
Merge branch 'feature/spin-taste_operator' of github.com:lattice/quda into feature/spin-taste_operator

AlexVaq committed:
Bug fixes

AlexVaq committed:
Added covDev for staggered

AlexVaq committed:
Bug fixes

AlexVaq committed:
Added covDev for staggered

AlexVaq committed:
Merge branch 'feature/spin-taste_operator' of github.com:lattice/quda into feature/spin-taste_operator

cpviolator committed:
Prevent global reduction in contraction kernels, populate a global length array with MPI local data, sum all data to node zero and return to user.

AlexVaq committed:
Fixed some minor potential bugs

Commit:
Merge branch 'feature/spin-taste_operator' of github.com:lattice/quda into feature/spin-taste_operator

Commit:
Dean's fix: Prevent global reduction in contraction kernels

Commit:
Merge branch 'develop' into try-merge
bring try-merge in sync with develop.

Commit:
Change vector_type to array in meson contractions.

Commit:
Merge branch 'feature/spin-taste_operator' into merge_st_corr_with_sp_tst

Commit:
Fix missing libmpi_cxx dependency when linking tests.

Commit:
Merge branch 'try-merge' into feature/staggered_correlators_gk
Bring this branch up to date with develop

Commit:
Merge branch 'develop' into feature/staggered_correlators_gk

cpviolator committed:
tweak output verbosity in correlation function

cpviolator committed:
Merge remote-tracking branch 'origin' into feature/hotqcd_correlators_gk_merged

cpviolator committed:
Rename vector_type to VectorType

cpviolator committed:
Use correct vector type

Commit:
Merge branch 'develop' into feature/staggered_correlators_gk

Commit:
Merge branch 'develop' into try-merge

Remaining issue with initialization of ReduceArg<contract_t> for summed contraction kernels?

Commit:
Merge branch 'develop' into try-merge

cpviolator committed:
develop merge

cpviolator committed:
Merge remote-tracking branch 'origin' into feature/hotqcd_correlators_gk_merged

cpviolator committed:
Develop merge

cpviolator committed:
Merge remote-tracking branch 'origin' into feature/hotqcd_correlators_gk_merged

cpviolator committed:
Merge remote-tracking branch 'origin' into feature/hotqcd_correlators_gk_merged

cpviolator committed:
develop merge

cpviolator committed:
Merge branch 'feature/reduction_comm_cleanup' into feature/hotqcd_correlators_gk_merged

cpviolator committed:
Update contrcation kernels to conform to new reduction classes

cpviolator committed:
fix minor bugs

cpviolator committed:
Mid debug checkpoint

cpviolator committed:
Revert to old return statement

cpviolator committed:
Merge branch 'feature/reduction_comm_cleanup' into feature/hotqcd_correlators_gk_merged

cpviolator committed:
Apply changes to contraction kernels to bring into line with newer reduction code

cpviolator committed:
Use sincos for trig in contractions

cpviolator committed:
Merge branch 'develop' into feature/hotqcd_correlators_gk_merged

cpviolator committed:
Merge remote-tracking branch 'origin' into feature/hotqcd_correlators_gk_merged

AlexVaq committed:
Remove debug printf statements and timers

AlexVaq committed:
Added flag to refresh the gauge field in qudaSpinTaste and qudaShift

AlexVaq committed:
Added guard for null fields

james-simone committed:
Alexei's fixes for arguments to ReduceArg.

james-simone committed:
Merge branch 'feature/staggered_correlators_gk' of github.com:lattice/quda into feature/staggered_correlators_gk

alexstrel committed:
Merge branch 'develop' into feature/staggered_correlators_gk

alexstrel committed:
Merge branch 'feature/hotqcd_correlators_gk_merged' into feature/staggered_correlators_gk

Commit:
fixed compilation errors

james-simone committed:
Add contractions on host to contractFT_test.

james-simone committed:
Fixed bug in three- and four-link staggered spin-taste operators in spinTasteQuda.

james-simone committed:
From hotfix/cmake_nvmath_libs. Fix to allow cmake to find the math libs
for CUDA bundled with nvhpc toolkits.

alexstrel committed:
Merge branch 'develop' into feature/staggered_correlators_gk

Commit:
corrected develop merge, a number of minor compilation fixes

maddyscientist committed:
Merge branch 'develop' into feature/staggered_correlators_gk

cpviolator committed:
Add make_4d_prop_quda source code file

james-simone committed:
Duplicate enum QudaContractGamma_s.

james-simone committed:
Restore QUDA_FFT_SYMM and QUDA_CONTRACT_TYPE enums.

james-simone committed:
Fixes for contaction kernels. File lib/interface_quda.cpp not compiling.

james-simone committed:
Restore StaggeredContract

cpviolator committed:
Merge branch 'develop' into not_a_feature_rather_a_hope/staggered_correlator_gk

cpviolator committed:
Edit out unwanted code, test suite compiles.

cpviolator committed:
Clean smearing of gauge field for wilson inversions

cpviolator committed:
Enable smearing for inverter, MG, and eigensolver. Need to think about how to handle mixed prec. Considering moving smearing call to loadGauge function.

james-simone committed:
Restore qudaContractFT to MILC.

james-simone committed:
Restore qudaContractFT signature.

cpviolator committed:
Move gauge smearing to loadGaugeQuda, encapsulate smearing into gauge parameter struct, add new gauge 'gaugeWorking', change logic for gauge aliasing

cpviolator committed:
Add boolean to gauge param to control smearing

cpviolator committed:
Differentiate fermion smearing operators for staggered and Wilson types, attempt to add in correletor_test, need to revise the call to contractFT

cpviolator committed:
Fuse wilson type fermion smearing routines

cpviolator committed:
fix gauge pointer logic in TwoLinK

cpviolator committed:
Update TwoLinkGaussianSmear interface function

cpviolator committed:
Merge in feature/quark-smearing.milc-interface, need to rework the freeUniqueGauge to comply with gaugeWorking

cpviolator committed:
Add QUDA_TWOLINK_LINKS link type, remove all gaugeWorking references, merge develop

cpviolator committed:
Add option to save gaugeTwoLink

cpviolator committed:
Fix typo in commit 1e1c5631d8c029dcde8f7ba3bf78d602cee2431e

james-simone committed:
Fixed typo

detar committed:
Change last arg (gauge refresh) of qudaSpinTaste to pointer

james-simone committed:
Merge branch 'hotfix/heavy-quark-restart' into not_a_feature_rather_a_hope/staggered_correlator_gk
The merge compiles, but is not tested.

james-simone committed:
Merge branch 'develop' into not_a_feature_rather_a_hope/staggered_correlator_gk

alexstrel committed:
Merge branch 'develop' into not_a_feature_rather_a_hope/staggered_correlator_gk

alexstrel committed:
Merge branch 'develop' into feature/staggered_correlators

alexstrel committed:
Merge branch 'develop' into feature/staggered_correlators

alexstrel committed:
Merge branch 'develop' into feature/staggered_correlators

alexstrel committed:
cleaned artifacts, added Jim to co-author list

maddyscientist commented:
Glad you got this pull request opened @alexstrel and @james-simone.  Once you get the CI tests to pass we can proceed to reviewing it.

maddyscientist commented on include/vector_io.h:
```diff
@@ -33,6 +33,12 @@ namespace quda
+       @brief Load propagator (12 vecs, Chroma compliant) from filename
+       @param[in] vecs The set of vectors to load
+    */
+    void loadProp(vector_ref<ColorSpinorField> &vecs);
```
In general, interfaces should take `cvector_ref`

alexstrel replied:
fixed in c2d6e23

maddyscientist commented on include/vector_io.h:
```diff
@@ -41,6 +47,20 @@ namespace quda
+       @param[in] vecs_low_prec The set of vectors with lower precision
+       @param[in] low_prec The low precsision value
+    */
+    void downPrec(cvector_ref<ColorSpinorField> &vecs_high_prec, vector_ref<ColorSpinorField> &vecs_low_prec,
```
Make second argument `cvector_ref`

alexstrel replied:
fixed in c2d6e23

maddyscientist commented on include/kernels/contraction.cuh:
```diff
@@ -8,14 +8,490 @@
+      // Color inner product: <\phi(x)_{\mu} | \phi(y)_{\nu}> ; The Bra is conjugated
+      complex<real> prop_prod = innerProduct(x, y, 0, 0);	
+
+      #if 0 // JNS
```
Can this debugging code be deleted?

alexstrel replied:
fixed in c2d6e23

maddyscientist commented on include/kernels/contraction.cuh:
```diff
@@ -8,14 +8,490 @@
+	}
+      }
+
+      // Debug
```
Delete?

alexstrel replied:
fixed in c2d6e23

maddyscientist commented on include/kernels/contraction.cuh:
```diff
@@ -8,14 +8,490 @@
+  using spinor_array = array<array<double, 2>, max_contract_results>;
+  using staggered_spinor_array = array<double, 2>; 
+
+  template <typename real, int nSpin = 4> class DRGammaMatrix {
```
Should this class definition be in its own header?

alexstrel replied:
@maddyscientist the class is used only in this file, so is there any other reason to isolate it?

maddyscientist replied:
I was suggesting this purely in the spirit of modularity: I figure the `DRGammaMatrix` class might be relevant in other kernels in the future, and it would make this kernel file cleaner to read.

alexstrel replied:
Addressed in 4d76007

maddyscientist commented on include/kernels/contraction.cuh:
```diff
@@ -8,14 +8,490 @@
+  public:
+    // Stores gamma matrix column index for non-zero complex value.
+    // This is shared by g5gm, gmg5.
+    int gm_i[nSpin*nSpin][nSpin] {};
```
Are the curly braces here to tell the default constructor to zero this array?  If so it should be written as `int gm_i[nSpin*nSpin][nSpin] = {}'`.  Also should there be equivalent for g5gm_z as well then?

alexstrel replied:
fixed in 84851f3

maddyscientist commented on include/contract_quda.h:
```diff
@@ -5,5 +5,11 @@
 namespace quda
 {
+  // called in interface_quda.cpp, defined in contract.cu
+  void contractSummedQuda(const ColorSpinorField &x, const ColorSpinorField &y, std::vector<Complex> &result,
```
Need to add doxygen here for this function.

alexstrel replied:
fixed with 28ffe81

maddyscientist commented on include/kernels/spin_taste.cuh:
```diff
@@ -0,0 +1,106 @@
+  }
+
+  template <typename Arg>
+  __device__ __host__ void applySign(int idx, int parity, const Arg &arg) {
```
Any reason not to just inline the body of this function into the `operator()` method of `SpinTastePhase`?

alexstrel replied:
fixed in f8208bd

maddyscientist commented on include/quda_milc_interface.h:
```diff
@@ -229,17 +242,60 @@ extern "C" {
+   * Apply the forward/backward/symmetric shift for the spin-taste opeartor. All fields
+   * passed and returned are host (CPU) field in MILC order.
+   *
+   * @param external_precision Precision of host fields passed to QUDA (2 - double, 1 - single)
```
Add `in/out` to these doxygen comments

alexstrel replied:
fixed in e0451ef

maddyscientist commented on include/quda_milc_interface.h:
```diff
@@ -229,17 +242,60 @@ extern "C" {
+   * Apply the forward/backward/symmetric shift for the spin-taste opeartor. All fields
+   * passed and returned are host (CPU) field in MILC order.
+   *
+   * @param external_precision Precision of host fields passed to QUDA (2 - double, 1 - single)
```
Add `in/out` to these doxygen comments

alexstrel replied:
fixed in e0451ef

maddyscientist commented on include/quda.h:
```diff
@@ -1280,6 +1287,35 @@ extern "C" {
 
+  /**
+   * Apply the covariant derivative.
+   * @param h_out  Result spinor field
```
Add `in/out` to these doxygen comments

alexstrel replied:
fixed in 0120341

maddyscientist commented on include/quda.h:
```diff
@@ -1280,6 +1287,35 @@ extern "C" {
+
+  /**
+   * Apply the covariant derivative.
+   * @param h_out  Result spinor field
```
Add `in/out` to these doxygen comments

alexstrel replied:
fixed in 0120341

maddyscientist commented on include/quda.h:
```diff
@@ -1280,6 +1287,35 @@ extern "C" {
+
+  /**
+   * Apply the spin-taste operator.
+   * @param h_out  Result spinor field
```
Add `in/out` to these doxygen comments

alexstrel replied:
fixed in 0120341

maddyscientist commented on include/quda.h:
```diff
@@ -1280,6 +1287,35 @@ extern "C" {
+   * @param dir    Direction of application
+   * @param param  Metadata for host and device storage
+   */
+  void covDevQuda(void *h_out, void *h_in, int dir, QudaInvertParam param);
```
The last argument should be passed by address (value) rather than by value.

alexstrel replied:
fixed in 92e6bf8

maddyscientist commented on include/quda.h:
```diff
@@ -1280,6 +1287,35 @@ extern "C" {
+   * @param sym    Apply forward=2, backward=2 or symmetric=3 shift
+   * @param param  Metadata for host and device storage
+   */
+  void shiftQuda(void *h_out, void *h_in, int dir, int sym, QudaInvertParam param);
```
Last argument should be passed as a pointer

alexstrel replied:
fixed in 92e6bf8

maddyscientist commented on include/quda.h:
```diff
@@ -1280,6 +1287,35 @@ extern "C" {
+   * @param taste  Taste gamma structure
+   * @param param  Metadata for host and device storage
+   */
+  void spinTasteQuda(void *h_out, void *h_in, int spin_, int taste, QudaInvertParam inv_param);
```
Last argument should be passed as a pointer

alexstrel replied:
fixed in 92e6bf8

maddyscientist commented on include/quda.h:
```diff
@@ -1670,6 +1733,14 @@ extern "C" {
                     const int *X);
 
+  /**
+   * DMH: to do
```
@cpviolator what's to be done here?

maddyscientist commented on lib/dslash_gamma_helper.cu:
```diff
@@ -10,21 +10,24 @@ namespace quda {
     void apply(const qudaStream_t &stream) {
       TuneParam tp = tuneLaunch(*this, getTuning(), getVerbosity());
-      launch<Gamma>(tp, stream, GammaArg<Float, nColor>(out, in, d));
+      if(proj == 0) launch<Gamma>(tp, stream, GammaArg<Float, nColor>(out, in, d, 0));
```
Maybe make the last argument to `GammaArg` default to 0 to avoid needing to specify 0 here and similarly on line 43 below?

alexstrel replied:
addressed in 0847503

maddyscientist commented on lib/spin_taste.cu:
```diff
@@ -0,0 +1,103 @@
+      TuneParam tp = tuneLaunch(*this, getTuning(), getVerbosity());
+      if (gamma == QUDA_SPIN_TASTE_G1) {
+        SpinTasteArg<Float, nColor, QUDA_SPIN_TASTE_G1> arg(out, in);
+        launch<SpinTastePhase>(tp, stream, arg);
```
Could this be made cleaner if you add
```c++
template <QudaSpinTasteGamma gamma> using Arg = SpinTasteArg<Float, nColor, gamma>;
```
above the apply function definition, and then each if / else case could written as
```c++
launch<SpinTastePhase>(tp, stream, Arg<Float, nColor, QUDA_SPIN_TASTE_G1>(out, in));
```
and it should be much more readable?

alexstrel replied:
fixed in da66551

maddyscientist commented on lib/vector_io.cpp:
```diff
@@ -79,6 +79,103 @@ namespace quda
+    if (vecs.size() != 12) errorQuda("Must have 12 vectors in propagator, passed %lu", vecs.size());
+    const int Nvec = vecs.size();
+    auto spinor_parity = v0.SuggestedParity();
+    if (getVerbosity() >= QUDA_SUMMARIZE) printfQuda("Start loading %04d vectors from %s\n", Nvec, filename.c_str());
```
Write as `logQuda`

alexstrel replied:
removed in c2d6e23

maddyscientist commented on lib/vector_io.cpp:
```diff
@@ -79,6 +79,103 @@ namespace quda
+      }
+    }
+
+    if (getVerbosity() >= QUDA_SUMMARIZE) printfQuda("Done loading vectors\n");
```
`logQuda`

alexstrel replied:
removed in c2d6e23

maddyscientist commented on lib/vector_io.cpp:
```diff
@@ -156,4 +253,115 @@ namespace quda
+      }
+    }
+
+    if (getVerbosity() >= QUDA_SUMMARIZE) printfQuda("Start saving %d vectors to %s\n", Nvec, filename.c_str());
```
`logQuda`

alexstrel replied:
removed in c2d6e23

maddyscientist commented on lib/vector_io.cpp:
```diff
@@ -156,4 +253,115 @@ namespace quda
+        for (int j = 0; j < Ls; j++) { V[i * Ls + j] = tmp[i].data<char*>() + j * stride; }
+      }
+
+      //write_propagator_field(filename.c_str(), &V[0], tmp[0]->Precision(), tmp[0]->X(), tmp[0]->SiteSubset(),
```
delete?

alexstrel replied:
removed in c2d6e23

maddyscientist commented on lib/vector_io.cpp:
```diff
@@ -156,4 +253,115 @@ namespace quda
+      errorQuda("Unexpected field dimension %d", vecs[0].Ndim());
+    }
+
+    if (getVerbosity() >= QUDA_SUMMARIZE) printfQuda("Done saving vectors\n");
```
`logQuda`

alexstrel replied:
removed in c2d6e23

maddyscientist commented on lib/vector_io.cpp:
```diff
@@ -156,4 +253,115 @@ namespace quda
+    if (getVerbosity() >= QUDA_SUMMARIZE) printfQuda("Done saving vectors\n");
+  }
+
+  void VectorIO::downPrec(cvector_ref<ColorSpinorField> &vecs_high_prec,
```
What's the plan for this function?

alexstrel replied:
removed in c2d6e23

maddyscientist commented on tests/CMakeLists.txt:
```diff
@@ -205,6 +211,11 @@ if(QUDA_CONTRACT)
   quda_checkbuildtest(contract_test QUDA_BUILD_ALL_TESTS)
   install(TARGETS contract_test ${QUDA_EXCLUDE_FROM_INSTALL} DESTINATION ${CMAKE_INSTALL_BINDIR})
+  
+  #add_executable(correlator_test correlator_test.cpp)
```
Why is this commented out?

alexstrel replied:
We don't need it, deleted in d62ea10bddc4bdf8621ec711c5adf628f13913ff

maddyscientist commented on tests/CMakeLists.txt:
```diff
@@ -123,6 +123,12 @@ if(QUDA_DIRAC_STAGGERED)
   install(TARGETS staggered_eigensolve_test ${QUDA_EXCLUDE_FROM_INSTALL} DESTINATION ${CMAKE_INSTALL_BINDIR})
 
+  if(QUDA_CONTRACT)
+    add_executable(contractFT_test contractFT_test.cpp)
```
We need a ctest instance that tests `contractFT_test`

alexstrel replied:
resolved in 4d43b5d46947a1a1b760b33ee5859d5d12bffc0a

maddyscientist commented on tests/contractFT_test.cpp:
```diff
@@ -0,0 +1,323 @@
+  initQuda(device_ordinal);
+  int X[4] = {xdim, ydim, zdim, tdim}; // local dims
+  setDims(X);
+  //cudaDeviceSetLimit(cudaLimitPrintfFifoSize,64*1024*1024); // DEBUG-JNS
```
delete?

alexstrel replied:
fixed in c2d6e23

maddyscientist commented on tests/contract_test.cpp:
```diff
@@ -135,15 +146,8 @@ int test(int contractionType, QudaPrecision test_prec)
-  default: errorQuda("Undefined contraction type %d\n", contractionType);
-  }
-
+  // We then compare the GPU result with a CPU refernce code.
```
refernce -> reference

alexstrel replied:
fixed in c2d6e23

maddyscientist commented on lib/contract.cu:
```diff
@@ -73,5 +245,5 @@ public:
   }
 #endif
-
+  //----------------------------------------------------------------------------
```
I don't think this line serves much purpose?

alexstrel replied:
removed in c2d6e23

maddyscientist commented on lib/covDev.cu:
```diff
@@ -37,7 +37,7 @@ namespace quda
       constexpr bool xpay = false;
       constexpr int nParity = 2;
-      Dslash::template instantiate<packShmem, nParity, xpay>(tp, stream);
+      Dslash::template instantiate<packStaggeredShmem, nParity, xpay>(tp, stream);
```
Does this change break nSpin = 4?

alexstrel replied:
@maddyscientist  I think you are right. While the code has been refocused towards staggered fermions, it still retains components that are implemented for nSpin = 4, e.g. 
https://github.com/lattice/quda/blob/feature/staggered_correlators/lib/covDev.cu#L138
etc. 
We could put if .. else ... block for each case , to take into account both versions. Any thoughts?

maddyscientist replied:
Yes, I think making if `if constexpr (Arg::nSpin) ....else...` would be the way to go.

alexstrel replied:
Actually, I was wrong regarding my previous comment since this operator is spin -agnostic (no spin related packing or projection). So I think we just keep it as is

maddyscientist replied:
Agree, I think this should be fine.  Marking as resolved.

maddyscientist commented on lib/milc_interface.cpp:
```diff
@@ -83,6 +83,7 @@ template <bool start> void inline qudamilc_called(const char *func, QudaVerbosit
     } else {
       printfQuda("QUDA_MILC_INTERFACE: %s (return) \n", func);
     }
+    fflush(stdout);
```
The `fflush` is redundant here as `printfQuda` include a call to `fflush`

alexstrel replied:
addressed in 61eb8fb2063699b3833a2bd1abcf361a3ca12803

maddyscientist commented on lib/milc_interface.cpp:
```diff
@@ -1260,14 +1261,104 @@ void qudaDslash(int external_precision, int quda_precision, QudaInvertArgs_t inv
+  }
+  invertParam.dslash_type = saveDslash;
+
+  if ((sym < 1) || (sym > 3)) {
```
Can `sym` vary between processes?  Should the subsequent `printf` call be `printfQuda`?

alexstrel replied:
addressed with 77c312753b2e9992d45564c1b3fdc4f9d3379dd5

maddyscientist commented on tests/su3_test.cpp:
```diff
@@ -22,9 +22,9 @@
-int gauge_smear_steps = 50;
-QudaGaugeSmearType gauge_smear_type = QUDA_GAUGE_SMEAR_STOUT;
-int measurement_interval = 5;
+//int gauge_smear_steps = 50;
```
Should these three lines be deleted instead of commented out?

alexstrel replied:
fixed in c2d6e23

maddyscientist commented on tests/utils/command_line_params.cpp:
```diff
@@ -1067,7 +1137,87 @@ void add_heatbath_option_group(std::shared_ptr<QUDAApp> quda_app)
   opgroup->add_option("--heatbath-warmup-steps", heatbath_warmup_steps,
-                      "Number of warmup steps in heatbath test (default 10)");
+                       "Number of warmup steps in heatbath test (default 10)");
+  // DMH
```
@cpviolator what's the status here?

maddyscientist commented on tests/utils/command_line_params.h:
```diff
@@ -7,9 +7,18 @@
 // for compatibility while porting - remove later
 extern void usage(char **);
 
+// Put this is quda_constants.h?
```
The answer to this question depends if these are universal constants or limited only to the tests?

weinbe2 replied:
Ping on this comment

maddyscientist commented on tests/utils/host_utils.cpp:
```diff
@@ -245,6 +245,7 @@ void constructWilsonTestSpinorParam(quda::ColorSpinorParam *cs_param, const Quda
   cs_param->location = QUDA_CPU_FIELD_LOCATION;
 }
 
```
Pointless addition?

alexstrel replied:
not sure what is wrong with the line?

maddyscientist commented on tests/utils/set_params.cpp:
```diff
@@ -177,6 +190,9 @@ void setInvertParam(QudaInvertParam &inv_param)
   }
 
+  // Gauge smear param
+  //setGaugeSmearParam(*inv_param.smear_param);
```
Delete?

alexstrel replied:
fixed in c2d6e23

maddyscientist commented on tests/utils/set_params.cpp:
```diff
@@ -970,6 +1002,9 @@ void setStaggeredInvertParam(QudaInvertParam &inv_param)
   inv_param.output_location = QUDA_CPU_FIELD_LOCATION;
 
+  // Gauge smear param
+  //setGaugeSmearParam(*inv_param.smear_param);
```
Delete?

alexstrel replied:
fixed in c2d6e23

maddyscientist commented on tests/contractFT_test.cpp:
```diff
@@ -0,0 +1,323 @@
+  // Perform GPU contraction.
+  contractFTQuda(spinorX, spinorY, &d_result, cType, (void*)(&cs_param), src_colors, X, source_position, n_mom, mom, fft_type);
+
+  #if 0
```
delete?

alexstrel replied:
fixed in c2d6e23

maddyscientist commented on lib/make_4d_prop_quda.cpp:
```diff
@@ -0,0 +1,88 @@
+    param4D.create = QUDA_REFERENCE_FIELD_CREATE;
+    L0.push_back(ColorSpinorField::Create(param4D));
+
+    std::vector<ColorSpinorField *> Lsm1;
```
Make `Lsm1` and `L0` std::vector<ColorSpinorField> instead of `std::vector<ColorSpinorField*>`

alexstrel replied:
fixed in 227a56d08f3f44e91b7332225df04cf296c35a68

maddyscientist commented on lib/interface_quda.cpp:
```diff
@@ -1842,6 +1872,457 @@ void dslashQuda(void *h_out, void *h_in, QudaInvertParam *inv_param, QudaParity
 
+void shiftQuda(void *h_out, void *h_in, int dir, int sym, QudaInvertParam inv_param)
+{
+  profileCovDev.TPSTART(QUDA_PROFILE_TOTAL);
```
The use of `profileCovDev` here should be cleaned up to be consistent with the "auto-profiling" we now have.  I can do this if it's easier.

alexstrel replied:
it's fixed in 32b92209db

maddyscientist commented on lib/interface_quda.cpp:
```diff
@@ -1842,6 +1872,457 @@ void dslashQuda(void *h_out, void *h_in, QudaInvertParam *inv_param, QudaParity
+  if (getVerbosity() >= QUDA_DEBUG_VERBOSE) printQudaInvertParam(&inv_param);
+
+  ColorSpinorParam cpuParam(h_in, inv_param, gauge.X(), false, inv_param.input_location);
+  ColorSpinorField *in_h = ColorSpinorField::Create(cpuParam);
```
Allocate on stack instead of heap

alexstrel replied:
fixed in 4fa8f65983905f955fa1dabad997b25e447653df

maddyscientist commented on lib/interface_quda.cpp:
```diff
@@ -1842,6 +1872,457 @@ void dslashQuda(void *h_out, void *h_in, QudaInvertParam *inv_param, QudaParity
+
+  cpuParam.v = h_out;
+  cpuParam.location = inv_param.output_location;
+  ColorSpinorField *out_h = ColorSpinorField::Create(cpuParam);
```
Move to stack

alexstrel replied:
fixed in 4fa8f65983905f955fa1dabad997b25e447653df

maddyscientist commented on lib/interface_quda.cpp:
```diff
@@ -1842,6 +1872,457 @@ void dslashQuda(void *h_out, void *h_in, QudaInvertParam *inv_param, QudaParity
+
+void spinTasteQuda(void *h_out, void *h_in, int spin_, int taste, QudaInvertParam inv_param)
+{
+  profileCovDev.TPSTART(QUDA_PROFILE_TOTAL);
```
Clean up profiling here too.

alexstrel replied:
addressed in 32b92209db

maddyscientist commented on lib/interface_quda.cpp:
```diff
@@ -1842,6 +1872,457 @@ void dslashQuda(void *h_out, void *h_in, QudaInvertParam *inv_param, QudaParity
+  if (getVerbosity() >= QUDA_DEBUG_VERBOSE) printQudaInvertParam(&inv_param);
+
+  ColorSpinorParam cpuParam(h_in, inv_param, gauge.X(), false, inv_param.input_location);
+  ColorSpinorField *in_h = ColorSpinorField::Create(cpuParam);
```
Move to stack

alexstrel replied:
fixed in 4fa8f65983905f955fa1dabad997b25e447653df

maddyscientist commented on lib/interface_quda.cpp:
```diff
@@ -1842,6 +1872,457 @@ void dslashQuda(void *h_out, void *h_in, QudaInvertParam *inv_param, QudaParity
+
+  cpuParam.v = h_out;
+  cpuParam.location = inv_param.output_location;
+  ColorSpinorField *out_h = ColorSpinorField::Create(cpuParam);
```
Move to stack

alexstrel replied:
fixed in 4fa8f65983905f955fa1dabad997b25e447653df

maddyscientist commented on lib/interface_quda.cpp:
```diff
@@ -1842,6 +1872,457 @@ void dslashQuda(void *h_out, void *h_in, QudaInvertParam *inv_param, QudaParity
+
+void covDevQuda(void *h_out, void *h_in, int dir, QudaInvertParam inv_param)
+{
+  profileCovDev.TPSTART(QUDA_PROFILE_TOTAL);
```
Profiling clean up

alexstrel replied:
addressed in 32b92209db

maddyscientist commented on lib/interface_quda.cpp:
```diff
@@ -1842,6 +1872,457 @@ void dslashQuda(void *h_out, void *h_in, QudaInvertParam *inv_param, QudaParity
+  if (getVerbosity() >= QUDA_DEBUG_VERBOSE) printQudaInvertParam(&inv_param);
+
+  ColorSpinorParam cpuParam(h_in, inv_param, gauge.X(), false, inv_param.input_location);
+  ColorSpinorField *in_h = ColorSpinorField::Create(cpuParam);
```
Move to stack

alexstrel replied:
fixed in 4fa8f65983905f955fa1dabad997b25e447653df

maddyscientist commented on lib/interface_quda.cpp:
```diff
@@ -1842,6 +1872,457 @@ void dslashQuda(void *h_out, void *h_in, QudaInvertParam *inv_param, QudaParity
+
+  cpuParam.v = h_out;
+  cpuParam.location = inv_param.output_location;
+  ColorSpinorField *out_h = ColorSpinorField::Create(cpuParam);
```
Move to stack

alexstrel replied:
fixed in 4fa8f65983905f955fa1dabad997b25e447653df

maddyscientist commented on lib/interface_quda.cpp:
```diff
@@ -4891,6 +5372,14 @@ void copyExtendedResidentGaugeQuda(void *resident_gauge)
   static_cast<GaugeField *>(resident_gauge)->copy(*extendedGaugeResident);
 }
 
+//void performGaussianSmearNStep(void *h_out, void *h_in, QudaInvertParam *param, const int n_steps, const double omega){
```
I see the legacy functions are still defined below.  Do these thin wrappers not work as expected yet, or can we delete the below and switch to these?

alexstrel replied:
removed in 6dfde4caaaa043937d1e3f38fffff6ea40752245

alexstrel replied:
additional object removed in f91077839edb17b7c925be3b72eae8769e687a14

maddyscientist commented on lib/interface_quda.cpp:
```diff
@@ -5311,3 +5912,188 @@ void gaugeObservablesQuda(QudaGaugeObservableParam *param)
   gaugeObservables(*gauge, *param);
 }
+
+#if 0
```
What purpose does this code serve (given it's all commented out)?

maddyscientist replied:
Delete this code

alexstrel replied:
the code was deleted in some commit

maddyscientist commented on lib/interface_quda.cpp:
```diff
@@ -5244,6 +5732,119 @@ int computeGaugeFixingFFTQuda(void *gauge, const unsigned int gauge_dir, const u
+
+  //FIXME can we merge the two propagators if they are the same to save mem?
+  // wrap CPU host side pointers
+  std::vector<ColorSpinorField*> h_prop1, h_prop2;
```
These can be moved to stack allocation to clean up this function.

alexstrel replied:
fixed in 4fa8f65983905f955fa1dabad997b25e447653df

maddyscientist commented on lib/interface_quda.cpp:
```diff
@@ -5244,6 +5732,119 @@ int computeGaugeFixingFFTQuda(void *gauge, const unsigned int gauge_dir, const u
+  cudaParam.gammaBasis = QUDA_DEGRAND_ROSSI_GAMMA_BASIS; // not relevant for staggered
+  cudaParam.setPrecision(cs_param->Precision(), cs_param->Precision(), true);
+  
+  std::vector<ColorSpinorField *> d_prop1, d_prop2;
```
Move to stack

alexstrel replied:
fixed in 4fa8f65983905f955fa1dabad997b25e447653df

maddyscientist reviewed (changes requested):
Visual review completed.  A few things to get the PR ready, but nothing that should be too hard.  Beyond the comments left a few outstanding issues are:
* Get the CI build matrix passing
* Add ctest instances to `tests/CMakeLists.txt` for the new tests and functionality added

alexstrel commented:
Thank you @maddyscientist for comments and suggestions.
what was the purpose of  the method at
https://github.com/lattice/quda/blob/feature/staggered_correlators/include/color_spinor_field.h#L789
It seems that the only place where it was supposed to be used is 
https://github.com/lattice/quda/blob/feature/staggered_correlators/lib/vector_io.cpp#L358 
but it is itself a commented code. 
Besides, it looks like that `downPrec` is now an abandoned function, so my suggestion is to delete it.

maddyscientist commented:
@alexstrel regarding `ColorSpinorField::CreateAlias`, it has been superseded by `ColorSpinorField::create_alias`, and as you point out it's not now used anywhere.  So I think we can just delete both `CreateAlias` and `downPrec` as you suggest.

alexstrel committed:
massive cleaning: removed artifacts

alexstrel committed:
corrected zero initialization for static arrays

alexstrel committed:
added inline keyword for applySign function

alexstrel committed:
corrected previous commit: moved function body of applySign into operator ()

alexstrel committed:
cleaned SpinTastePhase class

maddyscientist commented on lib/spin_taste.cu:
```diff
@@ -25,55 +25,41 @@ namespace quda {
 
     void apply(const qudaStream_t &stream)
     {
+      template <QudaSpinTasteGamma gamma> using Arg = SpinTasteArg<Float, nColor, gamma>;
```
This declaration needs to be outside of the `apply` function, e.g., on line 25.

alexstrel committed:
fixed commit 22b83a2

alexstrel committed:
Merge branch 'develop' into feature/staggered_correlators

alexstrel committed:
fixed compiler error (extra set of braces was messing)

alexstrel committed:
minor cleaning of the contraction.cuh, helper structures are now moved into gamma.cuh

alexstrel committed:
changed argument types in the interface functions (now are passed by pointers)

alexstrel committed:
using container of fields, instead of pointers

alexstrel committed:
moved objects to stack

alexstrel committed:
replaced spin_ with spin

alexstrel committed:
moved getSign sub-routine into operator()

alexstrel committed:
added a spin=1 support for covdev test

maddyscientist commented on tests/correlator_test.cpp:
```diff
@@ -0,0 +1,485 @@
+// QUDA headers
```
Delete this file - orthogonal to this PR

alexstrel replied:
Done in d62ea10bddc4bdf8621ec711c5adf628f13913ff

alexstrel committed:
now DR contraction kernel is isolated (but share uniform control structure), a couple of minor fixes

alexstrel committed:
removed correlator test

alexstrel committed:
Merge branch 'develop' into feature/staggered_correlators

alexstrel committed:
minor cleaning of the contraction kernel, add doxygen

maddyscientist committed:
Clean up profiling

maddyscientist committed:
Switch on CONTRACT, COVDEV and SMEAR_GAUSS_TWOLINK for cuda and rocm GitHub CI

detar committed:
For qudaSpinTaste and qudaShift, fixes three issues:
1. We never go beyond one link, so set the gauge parameters with a
   nullptr for the long link
2. Pass  the reloadGaugeField parameter by value. The caller manages
   resetting
3. Remove irrelevant conditions for reloading the gauge field.

detar committed:
In qudaSpinTaste, pass the reloadGaugeField parameter by value.

maddyscientist committed:
Merge branch 'not_a_feature_rather_a_hope/staggered_correlator_gk' into feature/staggered_correlators

maddyscientist committed:
Delete some chiral propagator code that's orthogonal to this PR

maddyscientist committed:
Merge remote-tracking branch 'origin' into feature/staggered_correlators

maddyscientist committed:
Update CMakeLists.txt

alexstrel committed:
Merge branch 'develop' into feature/staggered_correlators

alexstrel committed:
updated contraction test

alexstrel committed:
Merge branch 'develop' into feature/staggered_correlators

alexstrel committed:
corrected previous merge, note that this is related to gauge smearing interface

alexstrel committed:
fixed compilation issues

alexstrel committed:
source position is now set via CL option (--prop-source-position) , defaulted to {0,0,0,0}

alexstrel committed:
added missed parameter initialization

alexstrel committed:
minor cleaning, commented spin=4 block

maddyscientist commented:
@alexstrel have the contraction kernels been testing running on multiple GPU, when partitioning the XYZ dimension?  Looking at the code, I'm not sure that this case is properly supported.

alexstrel committed:
Merge branch 'develop' into feature/staggered_correlators

alexstrel committed:
fixed gtest, detected mem leak

alexstrel committed:
removed artifacts (redundant code led to memleak)

alexstrel committed:
updated covdev test application (needs some cleaning, though)

alexstrel committed:
Merge branch 'develop' into feature/staggered_correlators

alexstrel committed:
modified covdev test application: added covdev_mu option
this is an array of binary flags indicating which direction
to test. Default is (1,1,1,1) - testing all directions

alexstrel committed:
Merge branch 'develop' into feature/staggered_correlators

alexstrel committed:
Merge branch 'develop' into feature/staggered_correlators

alexstrel committed:
added error message when no direction is chosen for the covdev test

alexstrel committed:
Merge branch 'develop' into feature/staggered_correlators

alexstrel committed:
updated (corrected) doxygen comments for milc interface functions

alexstrel committed:
corrected doxygen comments in quda.h

alexstrel committed:
removed unnecessary argument (defaulted to zero)

alexstrel committed:
removed fflush

alexstrel committed:
replaced printf with printfQuda

alexstrel committed:
corrected previous commit with more strict error message

alexstrel committed:
Merge branch 'develop' into feature/staggered_correlators

alexstrel committed:
enabled ctest , removed contract_test.cpp (can be re-introduced later)

weinbe2 commented:
cscs-ci run

alexstrel committed:
Merge branch 'develop' into feature/staggered_correlators

alexstrel committed:
removed artifacts

alexstrel committed:
removed fermion_smear_transform object

maddyscientist committed:
contractFT_test needs the testing cml option group

maddyscientist committed:
Kernel cleanup

maddyscientist committed:
Apply clang format

maddyscientist reviewed (approved):
Thanks for the work that's been done on this @alexstrel @james-simone.  I think this is good to merge now.  There's a few performance and style issues that I would like to fix, but it shouldn't block this PR.  A great addition to QUDA.

maddyscientist committed:
Update ctest to run contractFT_test with testing enabled

weinbe2 commented on include/enum_quda_fortran.h:
```diff
@@ -23,6 +23,7 @@
 #define QUDA_MOMENTUM_LINKS 3
 #define QUDA_COARSE_LINKS   4
 #define QUDA_SMEARED_LINKS  5
+#define QUDA_TWOLINK_LINKS_LINKS 6
```
Should just be `QUDA_TWOLINK_LINKS`

alexstrel replied:
addressed in 514ce93

weinbe2 commented on include/enum_quda_fortran.h:
```diff
@@ -476,9 +477,36 @@
+
 #define QudaContractType integer(4)
-#define QUDA_CONTRACT_TYPE_OPEN ,
+#define QUDA_CONTRACT_TYPE_STAGGERED_FT_T ,
```
I recognize that this is just pattern-matching what was there before, but should these be defined to actual numerals instead of just ","?

alexstrel replied:
addressed in 8879b08e25fb129dc041a29c05df0ce732585d18

weinbe2 commented on include/kernels/contraction_old.cuh:
```diff
@@ -0,0 +1,524 @@
+#pragma once
```
Can we get rid of this file? I don't see it being included anywhere, but I only looked quickly.

alexstrel replied:
addressed dd7c47a5802d829e4cbd89f19149bc1cf86337cd

weinbe2 commented on tests/staggered_invert_test.cpp:
```diff
@@ -138,6 +138,13 @@ void init()
+    setGaugeSmearParam(smear_param);
+  }
+
+  if (!inv_multigrid) setStaggeredInvertParam(inv_param);
```
I think this can be removed?

alexstrel replied:
removed in 6925fbd407461ce9109a2238ff5caef85c84044c

weinbe2 commented on tests/utils/set_params.cpp:
```diff
@@ -977,6 +1004,9 @@ void setStaggeredInvertParam(QudaInvertParam &inv_param)
   inv_param.output_location = QUDA_CPU_FIELD_LOCATION;
 
+  // Gauge smear param
+  // setGaugeSmearParam(*inv_param.smear_param);
```
Delete unnecessary comment

alexstrel replied:
removed in 1b58a101d2e0765827a3e26792cac35d57313ac3

weinbe2 commented on tests/utils/command_line_params.h:
```diff
@@ -233,7 +355,9 @@ extern bool verify_results;
 extern bool low_mode_check;
 extern bool oblique_proj_check;
 extern double mass;
+extern quda::mass_array<double> mass_array;
```
I think `mass_array` and `kappa_array`, along with the corresponding options to set it on the command line, can be deleted, purely because they aren't being used anywhere. They're also somewhat redundant of `multishift_masses` in some regard... that said, I do like the logic in `add_massoption`, so maybe we want to make a note of it somewhere and apply it to `multishift_masses`, etc.

alexstrel replied:
I see `kappa_array` in `tests/host_reference/domain_wall_dslash_reference.cpp`. `mass_array` is indeed an artifact, will remove.

alexstrel replied:
fixed in 88a72635edcfd2a4e36f6bd9b9b71a6d34a59cc3

weinbe2 commented on tests/contractFT_test.cpp:
```diff
@@ -0,0 +1,264 @@
+
+  if (nspin == 1) {
+    faults = launch_contract_test<Float, 1, src_colors, n_mom>(cType, X, red_size, source_position, mom, fft_type);
+    //} else  if ( nspin == 4 ){
```
Should this be un-commented?

alexstrel replied:
No, since spin 4 case was not tested (activated) in this branch. We agreed that this can be done later, but I added a comment (TODO message)

weinbe2 commented on lib/interface_quda.cpp:
```diff
@@ -185,6 +189,15 @@ static TimeProfile profileGaugeObs("gaugeObservablesQuda");
 //!< Profiler for gaugeSmearQuda
 static TimeProfile profileGaugeSmear("gaugeSmearQuda");
 
+//!< Profiler for APEQuda
```
I think these can be removed, they look residual of some code that's been deleted.

alexstrel replied:
removed in 2ace7a2a010643023722cfc4c4b9758ea85e7073

weinbe2 commented on tests/covdev_test.cpp:
```diff
@@ -60,7 +64,7 @@ void init(int argc, char **argv)
   ColorSpinorParam csParam;
   csParam.nColor = nColor;
-  csParam.nSpin = 4;
+  csParam.nSpin = test_type == 0 ? 4 : 1; // use --test 1 for staggered case
```
Not a hill I'll die on, but we should put in an `errorQuda` if the test isn't 1 or 4, just because it's quick.

alexstrel replied:
addressed in 6323dc231273fc398ec7c6115fb551d519656705

weinbe2 commented on include/gamma.cuh:
```diff
@@ -291,4 +291,124 @@ namespace quda {
   };
 
+  // list of specialized structures used in the contraction kernels:
+  static constexpr int nspin = 4;
```
I think we should get rid of this define and just have the hard-coded magic "4" in the constexpr functions below, I don't like the risk of having a variable called `nspin` baked to be one specific thing or another.

alexstrel replied:
removed in 9dfbe57a5f2b5b5c98d76bf22a24137bb848816f

weinbe2 commented on include/quda.h:
```diff
@@ -1658,6 +1708,19 @@ extern "C" {
    */
   void performWuppertalnStep(void *h_out, void *h_in, QudaInvertParam *param, unsigned int n_steps, double alpha);
 
+  /**
```
We should make these "legacy" functions thin wrappers around `performFermionSmearQuda`. Nothing that needs to be done for this PR, but we should at least file an issue about it.

maddyscientist replied:
https://github.com/lattice/quda/issues/1474

alexstrel committed:
Merge branch 'develop' into feature/staggered_correlators

alexstrel committed:
corrected macro in enum_quda_fortran.h

alexstrel committed:
corrected enum_quda_fortran.h (added numerals)

alexstrel committed:
removed obsolete file

alexstrel committed:
removed artifacts

alexstrel committed:
removed commented snippet

alexstrel committed:
removed mass_array

alexstrel committed:
added a comment regarding spin=4 case

alexstrel committed:
removed artifacts

alexstrel committed:
added error message when test type is wrong

alexstrel committed:
removed static constexpr var to avoid possible unintended behavior

maddyscientist committed:
Apply clang-format

alexstrel committed:
Merge branch 'develop' into feature/staggered_correlators

maddyscientist committed:
contractFT -> contract_ft

maddyscientist committed:
Merge branch 'feature/staggered_correlators' of github.com:lattice/quda into feature/staggered_correlators

maddyscientist committed:
Remove dead code

maddyscientist committed:
Merge branch 'develop' of github.com:lattice/quda into feature/staggered_correlators

maddyscientist committed:
Contractions are now always enabled by default.  Remove QUDA_CONTRACT CMake flag (and various other deprecated cmake flags)

maddyscientist committed:
Tweak contact_ft_test to make it more robust

maddyscientist committed:
Switch cos/sin to cospi/sinpi in contraction kernel

maddyscientist committed:
Reinstate nSpin=4 Laplace operator (closes 475)

maddyscientist committed:
Apply clang format

maddyscientist committed:
CI fix
